api/admin/controller.py

```diff
@@ -815,3 +823,181 @@ def result_to_row(result):
+            if protocol:
+                collection, is_new = get_one_or_create(
+                    self._db, Collection, name=name, protocol=protocol
+                )
+            else:
+                return NO_PROTOCOL_FOR_NEW_COLLECTION
+
+        if protocol:
+            collection.protocol = protocol
+        if external_account_id:
+            collection.external_account_id = external_account_id
+        if url:
+            collection.url = url
+        if username:
+            collection.username = username
+        if password:
+            collection.password = password
+
+        for key, value in settings.items():
+            collection.setting(key).value = value
+        for setting in collection.settings:
+            if setting.key not in settings.keys():
+                self._db.delete(setting)
+
+
+        if collection.protocol == Collection.OPDS_IMPORT and not collection.url:
+            self._db.rollback()
+            return INCOMPLETE_COLLECTION_CONFIGURATION.detailed(_("An OPDS Import collection must have a url."))
+
+        if (collection.protocol == Collection.OVERDRIVE
+            and (not collection.username
+                 or not collection.password
+                 or not collection.external_account_id
+                 or not collection.setting("website_id").value)):
+            self._db.rollback()
+            return INCOMPLETE_COLLECTION_CONFIGURATION.detailed(_("An Overdrive collection must have username, password, external_account_id, and website_id."))
+
+        if (collection.protocol == Collection.BIBLIOTHECA
+            and (not collection.username
+                 or not collection.password
+                 or not collection.external_account_id)):
+            self._db.rollback()
+            return INCOMPLETE_COLLECTION_CONFIGURATION.detailed(_("A Bibliotheca collection must have username, password, and external_account_id."))
+
+        if (collection.protocol == Collection.AXIS_360
+            and (not collection.username
+                 or not collection.password
+                 or not collection.external_account_id)):
+            self._db.rollback()
+            return INCOMPLETE_COLLECTION_CONFIGURATION.detailed(_("An Axis 360 collection must have username, password, and external_account_id."))
+
+        if (collection.protocol == Collection.ONE_CLICK
+            and (not collection.username
+                 or not collection.password
+                 or not collection.external_account_id)):
+            self._db.rollback()
+            return INCOMPLETE_COLLECTION_CONFIGURATION.detailed(_("A OneClick collection must have username, password, and external_account_id."))
+
+        for short_name in libraries:
+            library = get_one(self._db, Library, short_name=short_name)
+            if not library:
+                return NO_SUCH_LIBRARY.detailed(_("%(library_short_name)s was added to the collection, but it does not exist.", library_short_name=short_name))
```

Comment:
This should make it clear that you _attempted_ to add a nonexistent library to the collection, not that it actually happened.